doc/design.md

```diff
@@ -0,0 +1,949 @@
+As part of the prototyping process we considered other solutions, in particular [dask](https://dask.org/).
+Dask could solve many problems such as overcoming the limitations of the fork-join approach as providing a unified solution for threading and multi-process parallelisation such as MPI.
+However, the prototyping experience indicated that it is not performant enough to completely replace threading for our typical workloads.
+For that reason, and for maximum simplicity, optimal performance requires multi-threading.
+
+##### Implementation
+
+In most cases a simple `#pragma omp parallel for` could do the job.
+We should probably limit the thread count depending on the size of data and the operation in question.
+For small data the threading overhead may be larger than the gains.
+For large data, many cases such as `operator+` are heavily limited by memory bandwidth and using more cores than necessary for reaching the peak bandwidth would be wasteful.
+
+
+### <a name="design-details-no-magic"></a>No magic
+
+##### Rationale
+
+Many algorithms in Mantid do "magic" under the hood:
+- Convert inputs if required. *Example: Conversion from point data in `Algorithms::ConvertUnits` and `Algorithms::Rebin`.*
+- Tolerate partially bad input data. *Example: Replacing special values in `Algorithms::SumSpectra`.*
+- Support doing two different things that are actually different. *Example: `Algorithms::Rebin` can rebin histogram data, bin event data, or set new bin edges in a workspace.*
+- Combine multiple processing steps. *Examples: Aligning bins in `Algorithms::ConvertUnits`. Loading and saving calibration files in `Algorithms::MaskDetectorsIf`.*
+
+Support for such features complicates the implementation and makes reasoning about code harder.
+It should therefore not be done in this library.
+- Operations should simply fail as early as possible.
+- Inputs are required to be cleaned up and converted to the correct format explicitly.
+- Processing steps should be kept as separate operations.
+
+##### Note
+
+There is nothing wrong with "magic" in higher-level operations.
+High level algorithms that simply work no matter what is thrown at them are likely still a requirement.
+We are merely banning it from *this* library, i.e., the lowest-level building block.
+
+
+### <a name="design-details-no-logging"></a>No logging
+
+##### Rationale
+
+As a corollary from the "no magic" strategy above, there will be very few occasions that require log messages.
+Logging can be implemented in higher-level algorithms, or potentially in some Python wrappers.
+Adding logging functionality would increase complexity and add extra dependencies.
+
+##### Workaround
+
+If required we could log to `stdout` using `std::cout`, which can be redirected and filtered.
+
+
+### <a name="design-details-exception-safety"></a>Operations have strong exception guarantee or clear data if only a basic exception guarantee can be given
```

Comment:
Hmm, you can only provide a strong exception guarantee by copying up front, that negates the purpose of doing it in-place in the first place.

I agree, we must not provide a weak guarantee with any of these operations. 

Clearing the data might be possible, but I feel there are more complexities to think about. i.e. what happens if an algorithm is modifying two datasets in-place at the same time and operation on one raises an exception.

Reply:
If all pre-conditions have been checked, why can't I provide a strong exception guarantee for in-place operations?

For example, `operator+=`: If I check that all dimensions match and units are ok, it can be in-place and guarantee that there is no exception?

Regarding you question on modifying two dataset at the same time: I am not sure how this is related to `Dataset`? Isn't that simply something the algorithm needs to handle in some way? If one thread fails, the other will just continue?

Reply:
My understanding of the strong guarantee is not that there will not be any exceptions, but in the presence of an exception, the state will be unchanged/un-corrupted https://en.wikipedia.org/wiki/Exception_safety. I think this differs from what you are saying about defensively trying to check inputs. I think this could be a matter of semantics, because I think for in-place operations your strategy is the only one that makes sense.

Regarding the other point, yes that would need to be handled at the algorithm level, and checking pre-conditions might be one way to do it.

Reply:
Yes as I said, there can be exceptions, but they will happen before an operation mutates the input. I do not see how that is different from the strong guarantee.

I am not sure why you need to check preconditions in the case of operating on two datasets in parallel, unless you are talking about and algorithm that modifies *two* inputs in distinct threads in-place, which sounds like a very special case.